276 add tdma communication scheme to the dotbot protocol

SaidAlvarado opened the pull request:
fixes #276

aabadie commented on doc/sphinx/drv.md:
```diff
@@ -26,6 +26,8 @@ _api/drv_pid
 _api/drv_protocol
 _api/drv_rgbled
 _api/drv_rgbled_pwm
+_api/tdma_client
+_api/tdma_server
```
```suggestion
_api/drv_tdma_client
_api/drv_tdma_server
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+
+typedef struct {
+    uint8_t  buffer[TDMA_CLIENT_RING_BUFFER_SIZE][PAYLOAD_MAX_LENGTH];  ///< arrays of radio messages waiting to be sent
+    uint32_t packet_length[TDMA_CLIENT_RING_BUFFER_SIZE];               ///< arrays of the lenght of the packets in the buffer
```
```suggestion
    uint32_t packet_length[TDMA_CLIENT_RING_BUFFER_SIZE];               ///< arrays of the length of the packets in the buffer
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+static tdma_client_vars_t _tdma_client_vars = { 0 };
+
+// Transform the ble mode into how many microseconds it takes to send a single byte.
+uint8_t ble_mode_to_byte_time[] = {
```
```suggestion
static const uint8_t ble_mode_to_byte_time[] = {
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+
+/**
+ * @brief Updates the RX and TX timings for the TDMA table.
+ *        DIrectly from an DB_PROTOCOL_TDMA_UPDATE_TABLE packet.
```
```suggestion
 *        Directly from an DB_PROTOCOL_TDMA_UPDATE_TABLE packet.
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+static void _message_rb_add(tdma_client_ring_buffer_t *rb, uint8_t data[PAYLOAD_MAX_LENGTH], uint8_t packet_length);
+
+/**
+ * @brief retreive the oldest element from the ring buffer for tdma captures
```
```suggestion
 * @brief retrieve the oldest element from the ring buffer for tdma captures
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+static void _tx_keep_alive_message(void);
+
+/**
+ * @brief register with the gatway to request a TDMA slot
```
```suggestion
 * @brief register with the gateway to request a TDMA slot
```

aabadie commented on drv/tdma_client.h:
```diff
@@ -0,0 +1,104 @@
+    uint32_t tx_duration;     ///< Duration of the DotBot's alloted frame.
+} tdma_client_table_t;
+
+typedef void (*tdma_client_cb_t)(uint8_t *packet, uint8_t length);  ///< Function pointer to the callback function called on packet receive
```
Wouldn't it make sense to also add rssi as a parameter to this callback function? Might be useful later

SaidAlvarado replied:
I agree, it would be useful.
I don't actually know how to do that, though.

I'll open a new issue for it, and do it later, I rather not delay this PR an further with extra features

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+typedef struct {
+    uint8_t  buffer[TDMA_CLIENT_RING_BUFFER_SIZE][PAYLOAD_MAX_LENGTH];  ///< arrays of radio messages waiting to be sent
+    uint32_t packet_length[TDMA_CLIENT_RING_BUFFER_SIZE];               ///< arrays of the lenght of the packets in the buffer
+    uint8_t  writeIndex;                                                ///< Index for next write
+    uint8_t  readIndex;                                                 ///< Index for next read
```
Use a consistent naming scheme. This repository uses lower snake case everwhere.
```suggestion
    uint8_t  write_index;                                                ///< Index for next write
    uint8_t  read_index;                                                 ///< Index for next read
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+}
+
+static bool _message_rb_get(tdma_client_ring_buffer_t *rb, uint8_t data[PAYLOAD_MAX_LENGTH], uint8_t *packet_length) {
+    if (rb->count <= 0) {
```
count is unsigned
```suggestion
    if (rb->count == 0) {
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+    // initialize variables
+    uint32_t start_tx_slot = db_timer_hf_now(TDMA_CLIENT_TIMER_HF);
+    uint8_t  length        = 0;
+    uint8_t  packet[PAYLOAD_MAX_LENGTH];
```
```suggestion
    uint8_t  packet[PAYLOAD_MAX_LENGTH] = { 0 };
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+    uint8_t  packet[PAYLOAD_MAX_LENGTH];
+    bool     packet_sent_flag = false;  ///< flag to keep track if a packet get sent during this function call
+
+    // check if there is something to send
+    if (_tdma_client_vars.tx_ring_buffer.count > 0) {
```
I would do it differently: if there's nothing to send, return false immediately.
Then no need to place the rest of the code in an if block and reenable radio rx afterwards (which is done even if nothing is sent in the current situation)

SaidAlvarado replied:
done!

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+static void _tx_keep_alive_message(void) {
+
+    db_protocol_header_to_buffer(_tdma_client_vars.radio_buffer, DB_BROADCAST_ADDRESS, TDMA_CLIENT_RADIO_APPLICATION, DB_PROTOCOL_ADVERTISEMENT);
+    size_t length = sizeof(protocol_header_t);
+    db_radio_disable();
+    db_radio_tx(_tdma_client_vars.radio_buffer, length);
```
```suggestion
    db_radio_disable();
    db_radio_tx(_tdma_client_vars.radio_buffer, sizeof(protocol_header_t));
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+static void _tx_tdma_register_message(void) {
+
+    db_protocol_header_to_buffer(_tdma_client_vars.radio_buffer, DB_BROADCAST_ADDRESS, TDMA_CLIENT_RADIO_APPLICATION, DB_PROTOCOL_ADVERTISEMENT);
+    size_t length = sizeof(protocol_header_t);
+    db_radio_disable();
+    db_radio_tx(_tdma_client_vars.radio_buffer, length);
```
```suggestion
    db_radio_disable();
    db_radio_tx(_tdma_client_vars.radio_buffer,  sizeof(protocol_header_t));
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+    }
+
+    // THe application type needs to be checked inside the application itself.
+    // We don't know it apriori
```
```suggestion
    // We don't know it a priori
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+                    _tdma_client_vars.tdma_client_table.rx_duration = frame_period;
+                }
+
+                // Also update the RX_duration, becaus ewe are working on ALWAYS_ON mode
```
```suggestion
                // Also update the RX_duration, because we are working on ALWAYS_ON mode
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+
+        return;
+
+    } else {
+        // Check if we just came from an idle period or an rx period
+        if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_WAIT) {
+            // Set the next interruption
+            db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, _tdma_client_vars.tdma_client_table.rx_duration, &timer_rx_interrupt);
+            // turn the radio ON
+            db_radio_rx();
+        } else if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_ON) {
+            // Set the next interruption
+            uint32_t delay = _tdma_client_vars.tdma_client_table.frame_duration - _tdma_client_vars.tdma_client_table.rx_duration;
+            db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, delay, &timer_rx_interrupt);
+            // turn the radio OFF
+            db_radio_disable();
+        }
+    }
```
No need for an else clause since you return in the if block
```suggestion
    }

    // Check if we just came from an idle period or an rx period
    if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_WAIT) {
        // Set the next interruption
        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, _tdma_client_vars.tdma_client_table.rx_duration, &timer_rx_interrupt);
        // turn the radio ON
        db_radio_rx();
    } else if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_ON) {
        // Set the next interruption
        uint32_t delay = _tdma_client_vars.tdma_client_table.frame_duration - _tdma_client_vars.tdma_client_table.rx_duration;
        db_timer_hf_set_oneshot_us(TDMA_CLIENT_TIMER_HF, TDMA_CLIENT_HF_TIMER_CC_RX, delay, &timer_rx_interrupt);
        // turn the radio OFF
       db_radio_disable();
    }
```

SaidAlvarado replied:
done!

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,522 @@
+            db_radio_disable();
+        }
+    }
+}
```
missing new line at end of file

aabadie reviewed (commented):
Partial review. Spotted a lot of typos. Will continue with the server implementation later

aabadie commented on projects/projects.emProject:
```diff
@@ -234,6 +234,45 @@
       </file>
     </folder>
   </project>
+  <project Name="01bsp_radio_onair_length_test">
```
The code of this application is missing it seems?

SaidAlvarado replied:
oh, I forgot to remove it. thanks!

aabadie reviewed (commented):
I'm done with my first review round.

SaidAlvarado commented on bsp/ipc.h:
```diff
@@ -65,12 +80,36 @@ typedef struct {
+    ipc_radio_pdu_t     rx_pdu;             ///< Received pdu
+} ipc_tdma_server_data_t;
+
+typedef struct {
```
I worry that a packed structure may miss-align the pointers to the internal variables and cause issue when I'm memcpy directly to and from  the ipc_shared_data structure.
For example in `db_tdma_server_get_client_info(tdma_table_entry_t *client, uint8_t client_id)`, [here](https://github.com/SaidAlvarado/DotBot-firmware/blob/276-add-tdma-communication-scheme-to-the-dotbot-protocol/drv/tdma_server/tdma_server.c#L258).

Though I might be wrong

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,691 @@
+    // check if there is something to send
+    if (rb->count > 0) {
+        // if yes, disable the radio.
+        db_radio_disable();
```
Oh, dear.
You just caught a gigantic bug.
This explains why I keep receiving a bunch of CRC errors that seems to be the messages leaving the radio getting echoed back.
Thanks for this!

SaidAlvarado replied:
Switching it to disabling the radio before every send

aabadie commented on bsp/ipc.h:
```diff
@@ -98,7 +137,14 @@ static inline void db_ipc_network_call(ipc_req_t req) {
-    while (!ipc_shared_data.net_ack) {}
+    while (!ipc_shared_data.net_ack) {
+        if (ipc_shared_data.req == DB_IPC_REQ_NONE) {
+            // Something went wrong and, the net-core deleted the request without fullfilling it.
```
```suggestion
            // Something went wrong and, the net-core deleted the request without fulfilling it.
```

SaidAlvarado replied:
fixed

aabadie commented on drv/protocol.h:
```diff
@@ -38,6 +38,8 @@ typedef enum {
     DB_PROTOCOL_CMD_XGO_ACTION     = 11,  ///< XGO action command
     DB_PROTOCOL_LH2_PROCESSED_DATA = 12,  ///< Lighthouse 2 data processed at the DotBot
+    DB_PROTOCOL_TDMA_UPDATE_TABLE  = 13,  ///< Receive new timings for the TDMA table
+    DB_PROTOCOL_TDMA_SYNC_FRAME    = 14,  ///< Sent by the gateway at the begining of a TDMA frame, if there is nothing else to send.
```
```suggestion
    DB_PROTOCOL_TDMA_SYNC_FRAME    = 14,  ///< Sent by the gateway at the beginning of a TDMA frame, if there is nothing else to send.
```

SaidAlvarado replied:
done

aabadie commented on drv/tdma_client.h:
```diff
@@ -0,0 +1,104 @@
+void db_tdma_client_tx(const uint8_t *packet, uint8_t length);
+
+/**
+ * @brief Ignore TDMA table and send all pending packets inmediatly
```
```suggestion
 * @brief Ignore TDMA table and send all pending packets immediately
```

SaidAlvarado replied:
fixed

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,520 @@
+    // Initialize Radio
+    db_radio_init(&tdma_client_callback, radio_mode);  // set the radio callback to our tdma catch function
+    db_radio_set_frequency(radio_freq);                // Pass through the rest of the arguments
+    db_radio_rx();                                     // start receving packets
```
```suggestion
    db_radio_rx();                                     // start receiving packets
```

SaidAlvarado replied:
fixed

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,520 @@
+ *
+ * This function will be called each time a radio packet is received.
+ * it will catch any TDMA related packet and update the timing table.
+ * All other types of packets are passed directly into the user-defined call-back
```
```suggestion
 * All other types of packets are passed directly into the user-defined callback
```

SaidAlvarado replied:
done

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,520 @@
+        return;
+    }
+
+    // THe application type needs to be checked inside the application itself.
```
```suggestion
    // The application type needs to be checked inside the application itself.
```

SaidAlvarado replied:
fixed

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,520 @@
+                // Protect against receiving garbage
+                if (frame_period > 0 && frame_period < 500000) {
+                    _tdma_client_vars.tdma_client_table.frame_duration = frame_period;
+                    // Also update the RX_duration, becaus ewe are working on ALWAYS_ON mode
```
```suggestion
                    // Also update the RX_duration, because we are working on ALWAYS_ON mode
```

SaidAlvarado replied:
fixed

aabadie commented on drv/tdma_server/tdma_server_nrf5340_app.c:
```diff
@@ -0,0 +1,131 @@
+//=========================== public ===========================================
+
+void db_tdma_server_init(tdma_server_cb_t callback, db_radio_ble_mode_t radio_mode, uint8_t radio_freq) {
+    db_hfclk_init();
+
+    // Disable all DCDC regulators (use LDO)
+    NRF_REGULATORS_S->VREGRADIO.DCDCEN = (REGULATORS_VREGRADIO_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGRADIO_DCDCEN_DCDCEN_Pos);
+    NRF_REGULATORS_S->VREGMAIN.DCDCEN  = (REGULATORS_VREGMAIN_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGMAIN_DCDCEN_DCDCEN_Pos);
+    NRF_REGULATORS_S->VREGH.DCDCEN     = (REGULATORS_VREGH_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGH_DCDCEN_DCDCEN_Pos);
+
+    // RADIO (address at 0x41008000 => periph ID is 8)
+    db_tz_enable_network_periph(NRF_NETWORK_PERIPH_ID_RADIO);
+    db_tz_enable_network_periph_dma(NRF_NETWORK_PERIPH_ID_RADIO);
+
+    // IPC (address at 0x41012000 => periph ID is 18)
+    db_tz_enable_network_periph(NRF_NETWORK_PERIPH_ID_IPC);
```
All this can be dropped, see #336 
```suggestion
```

aabadie commented on drv/tdma_client/tdma_client_nrf5340_app.c:
```diff
@@ -0,0 +1,129 @@
+//=========================== public ===========================================
+
+void db_tdma_client_init(tdma_client_cb_t callback, db_radio_ble_mode_t radio_mode, uint8_t radio_freq) {
+    db_hfclk_init();
+
+    // Disable all DCDC regulators (use LDO)
+    NRF_REGULATORS_S->VREGRADIO.DCDCEN = (REGULATORS_VREGRADIO_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGRADIO_DCDCEN_DCDCEN_Pos);
+    NRF_REGULATORS_S->VREGMAIN.DCDCEN  = (REGULATORS_VREGMAIN_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGMAIN_DCDCEN_DCDCEN_Pos);
+    NRF_REGULATORS_S->VREGH.DCDCEN     = (REGULATORS_VREGH_DCDCEN_DCDCEN_Disabled << REGULATORS_VREGH_DCDCEN_DCDCEN_Pos);
+
+    // RADIO (address at 0x41008000 => periph ID is 8)
+    db_tz_enable_network_periph(NRF_NETWORK_PERIPH_ID_RADIO);
+    db_tz_enable_network_periph_dma(NRF_NETWORK_PERIPH_ID_RADIO);
+
+    // IPC (address at 0x41012000 => periph ID is 18)
+    db_tz_enable_network_periph(NRF_NETWORK_PERIPH_ID_IPC);
```
All this is not needed (see #336)
```suggestion
```

aabadie commented on drv/tdma_server/tdma_server_nrf5340_app.c:
```diff
@@ -0,0 +1,131 @@
+    if (callback) {
+        _tdma_server_callback = callback;
+    }
+    db_timer_init(0);
+    NRF_P0->OUTCLR = 1 << 28;
+    db_timer_delay_ms(0, 5000);
+    NRF_P0->OUTCLR = 1 << 29;
```
This looks like leftover debugging code

aabadie commented on projects/03app_nrf5340_net/main.c:
```diff
@@ -83,24 +117,74 @@ int main(void) {
-            _req_received           = DB_IPC_REQ_NONE;
+            ipc_shared_data.net_ack       = true;
+            _nrf53_net_vars._req_received = DB_IPC_REQ_NONE;
+            ipc_shared_data.req           = DB_IPC_REQ_NONE;  // used in the app-core ipc.h to check that the request was properly fullfilled.
```
```suggestion
            ipc_shared_data.req           = DB_IPC_REQ_NONE;  // used in the app-core ipc.h to check that the request was properly fulfilled.
```

SaidAlvarado replied:
fixed

aabadie reviewed (commented):
Spotted some more typos and some code could also be removed (similar to #336)

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,691 @@
+#define TDMA_RADIO_APPLICATION SailBot
+#else
+#define TDMA_RADIO_APPLICATION DotBot
+#endif
```
It was to select which application_type_t to use in the automatic registration and sync messages sent by the TDMA system. This code attempts to auto select it based on the board definition. But it had no support for the freebot or lh2 mini mote

I rewrote it, now application_type_t it is a parameter in the tdma init functions.

in commit 42ff0032cdc3c9139e9e895756ec0780fcff653f

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,691 @@
+        }
+        // Update the last packet timer
+        _tdma_vars.last_tx_packet_ts = _tdma_vars.slot_start_ts;
+    }
```
Done

SaidAlvarado commented on projects/01drv_tdma_client/01drv_tdma_client.c:
```diff
@@ -0,0 +1,87 @@
+        db_timer_delay_ms(0, DELAY_MS);
+    }
+
+    // one last instruction, doesn't do anything, it's just to have a place to put a breakpoint.
+    __NOP();
```
done

SaidAlvarado commented on projects/01drv_tdma_server/01drv_tdma_server.c:
```diff
@@ -0,0 +1,92 @@
+        db_timer_delay_ms(0, DELAY_MS);
+    }
+
+    // one last instruction, doesn't do anything, it's just to have a place to put a breakpoint.
+    __NOP();
```
done

SaidAlvarado commented on projects/01drv_tdma_server/README.md:
```diff
@@ -0,0 +1 @@
+# LIS3MDL magnetometer sample application
```
done

SaidAlvarado commented on projects/01drv_tdma_client/README.md:
```diff
@@ -0,0 +1 @@
+# LIS3MDL magnetometer sample application
```
done

aabadie commented on projects/03app_dotbot/03app_dotbot.c:
```diff
@@ -342,10 +322,6 @@ static void _timeout_check(void) {
     }
 }
 
-static void _advertise(void) {
-    _dotbot_vars.advertize = true;
-}
```
Why remove the advertisement logic? It's used by the PyDotBot backend to track deadly robots so it's useful, even with TDMA.

SaidAlvarado replied:
The TDMA client takes care of advertisement internally, there is no need for the application to bother itself with it.

Here: https://github.com/DotBots/DotBot-firmware/blob/babec7bc1eaa5b792cd542f12260e2d493be647a/drv/tdma_client/tdma_client.c#L469-477

aabadie replied:
Ok, but only between the TMDA client and server. This advertizement packet is an application level logic between the robot and the pydotbot backend. To make sure that the robot is alive. I don't think they have the same purpose.

aabadie commented on projects/03app_lh2_mini_mote_app/03app_lh2_mini_mote_app.c:
```diff
@@ -47,7 +50,6 @@ static const gpio_t  _r_led_pin = { .port = DB_RGB_LED_PWM_RED_PORT, .pin = DB_R
 
 //=========================== prototypes =======================================
 
-static void _advertise(void);
```
Same comment about advertizement

SaidAlvarado replied:
same as above

aabadie commented on projects/03app_lh2_mini_mote_test/03app_lh2_mini_mote_test.c:
```diff
@@ -64,7 +65,6 @@ static const gpio_t  _b_led_pin = { .port = DB_RGB_LED_PWM_BLUE_PORT, .pin = DB_
 
 //=========================== prototypes =======================================
 
-static void _advertise(void);
```
Same comment about advertizement

SaidAlvarado replied:
same as above

aabadie reviewed (commented):
I found more typos and have a couple of new questions. Once they are addressed, I believe this PR will be very close to get in.

SaidAlvarado commented on bsp/ipc.h:
```diff
@@ -65,12 +81,38 @@ typedef struct {
-    ipc_req_t        req;        ///< IPC network request
-    ipc_radio_data_t radio;      ///< Radio shared data
-    ipc_rng_data_t   rng;        ///< Rng share data
+typedef struct {
```
I'll add this and check if it breaks something.

SaidAlvarado replied:
OK, can confirm it breaks compilation.

Segger fails compilation at this line: https://github.com/DotBots/DotBot-firmware/blob/babec7bc1eaa5b792cd542f12260e2d493be647a/projects/03app_nrf5340_net/main.c#L137

And complains about 

> taking address of packed member 'client_entry' of class or structure 'ipc_tdma_server_data_t' may result in an unaligned pointer value [-Werror,-Waddress-of-packed-member]

SaidAlvarado replied:
I'm un-packing the structure again

aabadie replied:
IMO this is a problem to not pack a struct shared by 2 separate cores. The main reason of the compilation error should be fixed instead.

SaidAlvarado replied:
Ok, I repacked this and fixed the compilation error.

SaidAlvarado commented on bsp/ipc.h:
```diff
@@ -65,12 +81,38 @@ typedef struct {
+    db_tdma_registration_state_t registration_state;  ///< db_tdma_client_get_status return value
+} ipc_tdma_client_data_t;
+
+typedef struct {
```
same as above

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+ *
+ * @param[in]    rb                     pointer to ring buffer structure
+ * @param[in]    max_tx_duration_us     max time available to send messages.
+ * @return                              true is a packet was sent, false if no packet was sent.
```
fixed

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+ *
+ * @param[in]    rb          pointer to ring buffer structure
+ * @param[out]   new_client  pointer to the variable where the new client ID will be saved.
+ * @return                   true if ID was successfully copied, false buffer is empty.
```
fixed

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+ *
+ * @param[in]    rb                  pointer to ring buffer structure
+ * @param[in]    max_tx_duration_us  max time available to send messages.
+ * @return                           true is a packet was sent, false if no packet was sent.
```
done

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+}
+
+static void _tx_registration_messages(uint64_t client) {
+    // Send the sync packet + tdma_slot parameters  to a particular client.
```
fixed

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+    // Save the timestamp start of the current slot, to ensure accurate computation of the end of the slot
+    _tdma_vars.slot_start_ts = db_timer_hf_now(TDMA_SERVER_TIMER_HF);
+
+    // If there are not enough clients, make sure the the minimum frame time is respected, even if is just filled with empty slots.
```
fixed

SaidAlvarado commented on projects/03app_dotbot/03app_dotbot.c:
```diff
@@ -28,15 +28,16 @@
 //=========================== defines ==========================================
 
+#define DB_RADIO_FREQ             (8U)      //< Set the frequency to 2408 MHz
+#define RADIO_APP                 (DotBot)  // DotBot Radio App
```
fixed

SaidAlvarado commented on projects/01drv_tdma_server/01drv_tdma_server.c:
```diff
@@ -0,0 +1,90 @@
+
+//=========================== defines ==========================================
+
+#define DELAY_MS   (1000)                // Wait between each send
+#define RADIO_FREQ (28)                  // Set the frequency to 2412 MHz
+#define RADIO_MODE (DB_RADIO_BLE_1MBit)  // Use BLE 1Mbit/s
+#define RADIO_APP  (DotBot)              // DotBot Radio App
```
fixed

SaidAlvarado commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,687 @@
+        return;
+    }
+
+    // Ignore messages that you sent yourself
+    if (header->src == _tdma_vars.device_id) {
+        return;
+    }
```
Not after you fixed the previous bug where I was transmitting without disabling the radio first.
I'm removing this check

SaidAlvarado commented on projects/01drv_tdma_server/01drv_tdma_server.c:
```diff
@@ -0,0 +1,90 @@
+static void radio_callback(uint8_t *packet, uint8_t length) {
+
+    if (packet[0] == length) {
+        __NOP();
```
This would mess up the other printf that display the status of the tdma server in a table of client IDs and frame information.
I'd rather keep it this way

aabadie commented on drv/tdma_server.h:
```diff
@@ -0,0 +1,108 @@
+
+/// Data type to store the info about a single TDMA Client
+typedef struct {
+    uint64_t client;       ///< ID of the client registered to in this time slot
```
The packing problem might come from using 64bit unsigned int on a 32bit aligned memory. Maybe changing this to `uint32_t client[2];` can help but then you have to memcpy somehow.

SaidAlvarado commented:
I'm observing an issue in which the Gateway is receiving an echo of every message it TX.
Maybe it happens as well in the DotBot.

I don't know why it happens, I'm disabling the radio before every send, as I'm supposed to.
I reinstated the check to discard messages that come from your same address.

It is a temporary solution, but after 3 hours of debugging radio states, I still have absolutely no idea of why this is occuring.
Maybe the radio ENABLE also enables the RX right before the send.

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,518 @@
+
+    // renable the Radio RX
+    if (_tdma_client_vars.rx_flag == DB_TDMA_CLIENT_RX_ON) {
+        db_radio_rx();
```
Could the echo problem come from this line? Normally, this function must be called once at the beginning or after a call to db_radio_disable. After db_radio_tx has completed, the radio automatically switch back to RX using shorts.

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,518 @@
+                _tdma_client_vars.rx_flag = DB_TDMA_CLIENT_RX_ON;
+
+                // Enable radio RX
+                db_radio_rx();
```
The radio is supposed to be in RX already, right?

SaidAlvarado replied:
Thanks for taking a look, let me check

SaidAlvarado commented:
#341 fixes the issue with the TX messages causing interruptions and looking like the message echoed back.
I implemented the fix here.
I works fine now, I removed the test that check that messages were not sent by yourself.

aabadie commented on drv/tdma_client.h:
```diff
@@ -0,0 +1,106 @@
+
+//=========================== defines ==========================================
+
+///< TDMA internal registrarion state
```
```suggestion
/// TDMA internal registrarion state
```

The `<` is used when the doxygen comment is place on the same line as the thing it document, at the end. When the doxygen comment is placed above, use `///`. Sorry to only notice that now.

aabadie commented on drv/tdma_client.h:
```diff
@@ -0,0 +1,106 @@
+    DB_TDMA_CLIENT_REGISTERED,    ///< the DotBot registered with the gateway
+} db_tdma_registration_state_t;
+
+///< TDMA internal RX state
```
```suggestion
/// TDMA internal RX state
```

aabadie commented on drv/tdma_client/tdma_client.c:
```diff
@@ -0,0 +1,513 @@
+
+static void tdma_client_callback(uint8_t *packet, uint8_t length);
+
+///< TDMA timer Interrupts
```
```suggestion
/// TDMA timer Interrupts
```

aabadie commented on drv/tdma_server/tdma_server.c:
```diff
@@ -0,0 +1,685 @@
+
+static void tdma_server_callback(uint8_t *packet, uint8_t length);
+
+///< TDMA timer Interrupts
```
```suggestion
/// TDMA timer Interrupts
```

aabadie commented on projects/03app_dotbot/03app_dotbot.c:
```diff
@@ -28,15 +28,17 @@
 #define DB_ADVERTIZEMENT_DELAY_MS (500U)   ///< 500ms delay between each advertizement packet sending
 #define DB_TIMEOUT_CHECK_DELAY_MS (200U)   ///< 200ms delay between each timeout delay check
 #define TIMEOUT_CHECK_DELAY_TICKS (17000)  ///< ~500 ms delay between packet received timeout checks
-#define DB_LH2_COUNTER_MASK       (0x07)   ///< Maximum number of lh2 iterations without value received
```
This seems like an unrelated change. Actually, I would just expect a drop in replacement of the radio calls with TDMA calls. All other changes should be reverted. Ideally, the diff should only contain changes related to radio -> tdma. So for instance the advertize callback should be place before the update_lh2 one, like before.

SaidAlvarado replied:
This is left over code from a time where the dotbot app waited for two LH sweeps before processing and sending the data. That define is part of a secondary advertisement mechanism that triggered and advertisement after a certain amount of LH2 pulses arrived. The LH2 code no longer works that way, and this code serves no purpose.

While I reinstated app-level advertisement, I noticed this piece of unused code.
I think it is reasonable to remove it here.

aabadie commented on projects/03app_lh2_mini_mote_app/03app_lh2_mini_mote_app.c:
```diff
@@ -25,19 +25,22 @@
     uint32_t ts_last_packet_received;            ///< Last timestamp in microseconds a control packet was received
     db_lh2_t lh2;                                ///< LH2 device descriptor
     uint8_t  radio_buffer[DB_BUFFER_MAX_BYTES];  ///< Internal buffer that contains the command to send (from buttons)
-    bool     advertize;                          ///< Whether an advertize packet should be sent
```
The advertize feature removal has not been reverted in this application.

aabadie commented on projects/03app_lh2_mini_mote_test/03app_lh2_mini_mote_test.c:
```diff
@@ -49,7 +51,6 @@ typedef struct {
     uint32_t    ts_last_packet_received;            ///< Last timestamp in microseconds a control packet was received
     db_lh2_t    lh2;                                ///< LH2 device descriptor
     uint8_t     radio_buffer[DB_BUFFER_MAX_BYTES];  ///< Internal buffer that contains the command to send (from buttons)
-    bool        advertize;                          ///< Whether an advertize packet should be sent
```
Same comment about advertize

aabadie reviewed (commented):
I am also realizing that 03app_sailbot, 03app_xgo and 03app_freebot applications have to be updated to use tdma now.
